src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferLengthCountIsEmptyOverAnyAnalyzer.cs

```diff
@@ -0,0 +1,142 @@
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(PreferLengthOverAnyDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false
+        );
+
+        internal static readonly DiagnosticDescriptor CountDescriptor = DiagnosticDescriptorHelper.Create(
+            CountId,
+            CreateLocalizableResourceString(nameof(PreferCountOverAnyTitle)),
+            CreateLocalizableResourceString(nameof(PreferCountOverAnyMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(PreferCountOverAnyDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false
+        );
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterOperationAction(OnInvocationAnalysis, OperationKind.Invocation);
+        }
+
+        private static void OnInvocationAnalysis(OperationAnalysisContext context)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+            var originalMethod = invocation.TargetMethod.OriginalDefinition;
+            if (IsEligibleAnyMethod(originalMethod))
+            {
+                // The collection will be passed as the instance in VB and as an argument in C#. 
+                var firstArgument = invocation.Instance ?? invocation.Arguments[0].Value;
+                var type = (firstArgument as IConversionOperation)?.Operand.Type ?? firstArgument.Type;
+                if (HasEligibleIsEmptyProperty(type))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(IsEmptyDescriptor, invocation.Syntax.GetLocation()));
+
+                    return;
+                }
+
+                if (HasEligibleLengthProperty(type))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(LengthDescriptor, invocation.Syntax.GetLocation()));
+
+                    return;
+                }
+
+                if (HasEligibleCountProperty(type))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(CountDescriptor, invocation.Syntax.GetLocation()));
+
+                    return;
+                }
+            }
+        }
+
+        private static bool IsEligibleAnyMethod(IMethodSymbol method)
+        {
+            return method.Name == AnyText
+                   && method.ReturnType.SpecialType == SpecialType.System_Boolean
+                   && method.Language == LanguageNames.CSharp && method.Parameters.Length == 1
+                   || (method.Language == LanguageNames.VisualBasic && (method.Parameters.Length == 0 || method.Parameters.Length == 1 && method.Parameters[0].Name == SourceParameterName));
```

Comment:
Something looks smelly here. In VB, you don't check the method name and return type at all.
In C#, you don't check the parameter name.

In general, I think this should be a symbol comparison.

Reply:
> In general, I think this should be a symbol comparison.

In case we want to restrict to a specific symbol like `Enumerable.Any()`

Reply:
Turns out we want to flag for any type, not only IEnumerable.Any()